[Java][jaxrs-spec] Add option to use jakarta imports / namespace

thmue opened the pull request:
Support Jakarta EE in jaxrs-spec.
partially addresses:
https://github.com/OpenAPITools/openapi-generator/issues/13124  
https://github.com/OpenAPITools/openapi-generator/issues/6881




To enable jakarta imports, set the additional property useJakartaEE to true.
If the option is false the legacy javax.* namespace is imported.


example:
java -jar ./openapi-generator-cli.jar generate -g jaxrs-spec --additional-properties=useJakartaEE=true

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

thmue commented:
i updated the author info, all build tests passed on circleCI

jorgerod commented:
Hi

It would be nice to fix all Java clients, not only jaxrs-spec. What do you think @thmue @wing328 ?

thmue commented:
Hi :-)
i think in principle all implementations which depend on jakartaEE could benefit from adding this flag.  

right now, under ressources/JavaJaxRS are several independent (?) implementations. There is also one at the root level of this directory (is that still relevant?). Would it be necessary or desired to unify them somehow? (if yes, how?)

shamus13 commented:
I am very interested in getting this PR merged, as i am working on bumping our Tomcat / JBoss webapps to Jakarta EE 9 spec. 
The PR looks good to go to me, if the merge conflicts get resolved, but what are your thoughts on this @thmue @wing328 ?

shamus13 commented:
@thmue i have looked into the build problem a bit, and i am almost certain, that it is not due to this PR at all.

I believe, the problem is with the pom template(s) missing a version field for the spring-boot-maven-plugin, which causes maven to pick 3.0.0, which again has transitive dependencies built for java 17.

Try adding a version field in the template; modules/openapi-generator/src/main/resources/java-camel-server/pom.mustache

Something like this;
...
                 <groupId>org.springframework.boot</groupId>
                 <artifactId>spring-boot-maven-plugin</artifactId>
                <version>${org.springframework.boot.version}</version>
...

This module is not the only place, i think, this problem exists.

thmue commented:
Thank you @shamus13 for debugging this issue. I added the spring version to the pom.xml (samples/server/petstore/java-camel/pom.xml )

Additionally, i added the fix also to the pom.mustache, however, this file does not influence the build

@wing328 would it be possible to get this merged during the 6.3.0 merge window?

shamus13 commented:
Glad to be of help. Anything to get this merged. I really hope, @wing328 can find the time to have a look at this again.

davidgamero commented:
just ran into this on #14319 while working on typescript experimental, hoping it gets merged soon!

shamus13 commented:
It's been a while now, perhaps one the java team could have a look and help speed things along?

@bbdouglas @sreeshas @jfiala @lukoyanov @cbornet @jeff9finger @karismann @Zomzog @lwlee2608

wing328 commented on bin/configs/jaxrs-spec-ee.yaml:
```diff
@@ -0,0 +1,11 @@
+generatorName: jaxrs-spec
+outputDir: samples/server/petstore/jaxrs-spec
```
i think the output folder cannot be the same. please change it to `samples/server/petstore/jaxrs-spec-ee` instead

wing328 replied:
please also update [.github/workflows/samples-jaxrs.yaml](https://github.com/OpenAPITools/openapi-generator/blob/037b41e1739ad7071afb875388b09bbd317ae67c/.github/workflows/samples-jaxrs.yaml) with the new output directory

wing328 commented on bin/configs/jaxrs-spec-ee.yaml:
```diff
@@ -0,0 +1,11 @@
+generatorName: jaxrs-spec
+outputDir: samples/server/petstore/jaxrs-spec
+inputSpec: modules/openapi-generator/src/test/resources/2_0/petstore-with-fake-endpoints-models-for-testing.yaml
```
please use modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml (3.0 spec) for testing as we're moving away from 2.0 spec.

wing328 commented:
please also run `./bin/utils/ensure-up-to-date` to update the doc and the samples.

wing328 commented:
@thmue looks like some github actions still failed: https://github.com/OpenAPITools/openapi-generator/actions/runs/3788726406/jobs/6441805752. Can you please take a look?

The workflow uses JDK8. Do we need to test it with JDK11 instead? 

(there's another workflow to test with JDK11)

thmue commented:
it seems that recent versions of jakarta.ws.rs.* (3.1) are only build for 11+

see minimum version : https://jakarta.ee/specifications/restful-ws/3.1/

thmue commented:
ok the builds pass now :-)

wing328 commented:
https://github.com/OpenAPITools/openapi-generator/actions/runs/3788849502/jobs/6446526010 still failed. please take another look.

Before this PR, the build of `samples/serer/petstore/jaxrs-spec) is perfectly fine with JDK8 tought. 

I'll another look later this week to see if I find out why.

thmue commented:
After generating the sample, three generated model classes
- `EnumTest.java`
- `NullableClass.java`
-` HealthCheckResult.java`

try to import:

`org.openapitools.jackson.nullable`

Of course this package does not exist. I don't know where / why this import is generated.

wing328 commented:
can you please resolve the merge conflicts when you've time? thanks.

borsch commented:
@thmue `useJakartaEe` option already added to all java generators which extends `AbstractJavaCodegen`. Could you please try 6.3.0 version?

More information can be found here https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/jaxrs-spec.md

jorgerod commented:
> @thmue `useJakartaEe` option already added to all java generators which extends `AbstractJavaCodegen`. Could you please try 6.3.0 version?
> 
> More information can be found here https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/jaxrs-spec.md

I use `useJakartaEe` option for RestTemplate and for WebClient and it works ok

thmue commented:
with https://github.com/OpenAPITools/openapi-generator/pull/14310  by @borsch merged, there is no need for this PR anymore. I ll therefore close it, thank you @wing328 for your support.
